db/repair_bucket_test.go

```diff
@@ -0,0 +1,114 @@
+package db
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"testing"
+
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbaselabs/go.assert"
+)
+
+const (
+	docIdProblematicRevTree = "docIdProblematicRevTree"
+)
+
+func testBucketWithViewsAndBrokenDoc() (bucket base.Bucket, numDocs int) {
+
+	numDocsAdded := 0
+	bucket = testBucket()
+	log.Printf("installing views")
+	installViews(bucket)
+	log.Printf("installed views")
+
+	// Add a harmless doc
+	testSyncData := syncData{}
+	bucket.Add("foo", 0, map[string]interface{}{"foo": "bar", "_sync": testSyncData})
+	numDocsAdded++
+
+	// Add a doc that should be repaired
+	rawDoc, err := unmarshalDocument(docIdProblematicRevTree, []byte(testdocProblematicRevTree))
+	if err != nil {
+		panic(fmt.Sprintf("Error unmarshalling doc: %v", err))
+	}
+	bucket.Add(docIdProblematicRevTree, 0, rawDoc)
+	numDocsAdded++
+
+	return bucket, numDocsAdded
+
+}
+
+func TestRepairBucket(t *testing.T) {
+
+	base.EnableLogKey("CRUD")
+
+	bucket, _ := testBucketWithViewsAndBrokenDoc()
+
+	repairJob := func(docId string, originalCBDoc []byte) (transformedCBDoc []byte, transformed bool, err error) {
+		log.Printf("repairJob called back")
+		return nil, true, nil
+	}
+	repairBucket := NewRepairBucket(bucket).
+		SetDryRun(true).
+		AddRepairJob(repairJob)
+
+	repairedDocs, err := repairBucket.RepairBucket()
+
+	assertNoError(t, err, fmt.Sprintf("Unexpected error: %v", err))
+
+	// Both docs will be repaired due to the repairJob function that indiscriminately repairs all docs
+	assert.True(t, len(repairedDocs) == 2)
+
+}
+
+func TestRepairBucketRevTreeCycles(t *testing.T) {
+
+	base.EnableLogKey("CRUD")
```

Comment:
Run this with two docs w/ cycles (can be the same body) to ensure everything is solid with loops/recursion.

Reply:
Fixed